### PR TITLE
refactor!: Particle hypothesis for track parameters

### PR DESCRIPTION
**Pull request by andiwand**
In this PR I make the particle hypothesis part of the track parameters. This allows us to consistently use the same hypothesis across various parts of our code. Before that the particle hypothesis was a bit scattered in the propagator and stepper.

replaces https://github.com/acts-project/acts/pull/2181 after splitting it up

blocked by
- https://github.com/acts-project/acts/pull/2366
- https://github.com/acts-project/acts/pull/2396
- https://github.com/acts-project/acts/pull/2398
- https://github.com/acts-project/acts/pull/2397

### CODE REVIEW

File: Tests/UnitTests/Core/TrackFitting/Gx2fTests.cpp [hunk 1]
```diff
@@ -62,7 +62,8 @@ Acts::CurvilinearTrackParameters makeParameters(
   Acts::BoundSquareMatrix cov = stddev.cwiseProduct(stddev).asDiagonal();
   // define a track in the transverse plane along x
   Acts::Vector4 mPos4(x, y, z, w);
-  return Acts::CurvilinearTrackParameters(mPos4, phi, theta, p, q, cov);
+  return Acts::CurvilinearTrackParameters(mPos4, phi, theta, q / p, cov,
+                                          Acts::ParticleHypothesis::pion());
```
Comment (benjaminhuth) [inline Tests/UnitTests/Core/TrackFitting/Gx2fTests.cpp:66]: @AJPfleger just to inform you about that new default hiere

File: Core/include/Acts/Vertexing/HelicalTrackLinearizer.ipp [hunk 2]
```diff
@@ -79,12 +79,9 @@ Acts::Result<Acts::LinearizedTrack> Acts::
...
   ActsScalar qOvP = paramsAtPCA(BoundIndices::eBoundQOverP);
 
-  // Get mass hypothesis from propagator options
-  ActsScalar m0 = pOptions.mass;
-  // Assume unit charge
-  // TODO: Get charge hypothesis from propagator options once they are included
-  // there.
-  ActsScalar p = std::abs(1. / qOvP);
+  ActsScalar m0 = params.particleHypothesis().mass();
+  ActsScalar p = std::abs(params.particleHypothesis().absoluteCharge() / qOvP);
```
Comment (benjaminhuth) [inline Core/include/Acts/Vertexing/HelicalTrackLinearizer.ipp:83]: @felix-russo Just to inform you that this in principle can change behaviour, but should resolve the `TODO` that was there correctly

Comment (felix-russo) [reply Core/include/Acts/Vertexing/HelicalTrackLinearizer.ipp:83]: Looks great! 👍

File: Core/include/Acts/EventData/TrackProxy.hpp [hunk 3]
```diff
@@ -362,10 +357,31 @@ class TrackProxy {
...
+  /// @return The absolute track momentum
+  ActsScalar charge() const {
+    return particleHypothesis().qFromQOP(qOverP());
+  }
+
   /// Get the absolute momentum of the tack
   /// @return The absolute track momentum
   ActsScalar absoluteMomentum() const {
-    return SinglyCharged{}.extractMomentum(qOverP());
+    return particleHypothesis().extractMomentum(qOverP());
```
Comment (paulgessinger) [inline Core/include/Acts/EventData/TrackProxy.hpp:384]: And in this case of temporary access this becomes a bit hairy since we might be recreating these objects quite frequently I suppose.

File: Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp [hunk 4]
```diff
@@ -77,8 +77,12 @@ struct GenericDenseEnvironmentExtension {
             typename navigator_t>
   int bid(const propagator_state_t& state, const stepper_t& stepper,
           const navigator_t& navigator) const {
+    auto particleHypothesis = stepper.particleHypothesis(state.stepping);
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp:80]: Copy?

File: Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp [hunk 5]
```diff
@@ -116,28 +120,34 @@ struct GenericDenseEnvironmentExtension {
          const navigator_t& navigator, ThisVector3& knew, const Vector3& bField,
          std::array<Scalar, 4>& kQoP, const int i = 0, const double h = 0.,
          const ThisVector3& kprev = ThisVector3::Zero()) {
+    // using because of autodiff
+    using std::hypot;
+
     auto q = stepper.charge(state.stepping);
+    auto particleHypothesis = stepper.particleHypothesis(state.stepping);
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp:127]: copy?

File: Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp [hunk 6]
```diff
@@ -171,6 +181,12 @@ struct GenericDenseEnvironmentExtension {
             typename navigator_t>
   bool finalize(propagator_state_t& state, const stepper_t& stepper,
                 const navigator_t& /*navigator*/, const double h) const {
+    // using because of autodiff
+    using std::hypot;
+
+    auto particleHypothesis = stepper.particleHypothesis(state.stepping);
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp:187]: copy?

File: Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp [hunk 7]
```diff
@@ -383,23 +394,32 @@ struct GenericDenseEnvironmentExtension {
...
-  template <typename propagator_state_t>
-  void initializeEnergyLoss(const propagator_state_t& state) {
-    energy[0] = hypot(initialMomentum, state.options.mass);
+  template <typename propagator_state_t, typename stepper_t>
+  void initializeEnergyLoss(const propagator_state_t& state,
+                            const stepper_t& stepper) {
+    // using because of autodiff
+    using std::hypot;
+
+    auto particleHypothesis = stepper.particleHypothesis(state.stepping);
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/detail/GenericDenseEnvironmentExtension.hpp:406]: copy?

File: Core/include/Acts/Vertexing/AMVFInfo.hpp
Comment (paulgessinger) [inline Core/include/Acts/Vertexing/AMVFInfo.hpp:1]: Revert maybe?

File: Core/src/Material/SurfaceMaterialMapper.cpp [hunk 8]
```diff
@@ -231,9 +232,10 @@ void Acts::SurfaceMaterialMapper::mapInteraction(
   std::map<GeometryIdentifier, unsigned int> assignedMaterial;
   using VectorHelpers::makeVector4;
   // Neutral curvilinear parameters
-  NeutralCurvilinearTrackParameters start(makeVector4(mTrack.first.first, 0),
-                                          mTrack.first.second,
-                                          1 / mTrack.first.second.norm());
+  NeutralCurvilinearTrackParameters start(
+      makeVector4(mTrack.first.first, 0), mTrack.first.second,
+      1 / mTrack.first.second.norm(), std::nullopt,
+      NeutralParticleHypothesis::photon());
```
Comment (paulgessinger) [inline Core/src/Material/SurfaceMaterialMapper.cpp:238]: Are these actually photons? Might be geantinos I suppose. Unsure if that makes a difference downstream.

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 9]
```diff
@@ -365,9 +366,10 @@ void Acts::VolumeMaterialMapper::mapMaterialTrack(
   using VectorHelpers::makeVector4;
 
   // Neutral curvilinear parameters
-  NeutralCurvilinearTrackParameters start(makeVector4(mTrack.first.first, 0),
-                                          mTrack.first.second,
-                                          1 / mTrack.first.second.norm());
+  NeutralCurvilinearTrackParameters start(
+      makeVector4(mTrack.first.first, 0), mTrack.first.second,
+      1 / mTrack.first.second.norm(), std::nullopt,
+      NeutralParticleHypothesis::photon());
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:372]: Same here, geantinos?

File: Examples/Algorithms/Propagation/include/ActsExamples/Propagation/PropagationAlgorithm.hpp [hunk 10]
```diff
@@ -107,6 +108,9 @@ class PropagationAlgorithm : public IAlgorithm {
     /// pt range
     std::pair<double, double> ptRange = {100 * Acts::UnitConstants::MeV,
                                          100 * Acts::UnitConstants::GeV};
+    /// particle hypothesis
+    Acts::ParticleHypothesis particleHypothesis =
+        Acts::ParticleHypothesis::pion();
```
Comment (paulgessinger) [inline Examples/Algorithms/Propagation/include/ActsExamples/Propagation/PropagationAlgorithm.hpp:113]: This was the implicit default before, right? Can we set this from python?

File: Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/TrackParameterSelector.cpp [hunk 11]
```diff
@@ -42,8 +42,9 @@ ActsExamples::ProcessCode ActsExamples::TrackParameterSelector::execute(
   auto isValidTrack = [&](const auto& trk) {
     const auto theta = trk.template get<Acts::eBoundTheta>();
     const auto eta = -std::log(std::tan(theta / 2));
+    const auto pT = trk.transverseMomentum();
     // define charge selection
-    return within(trk.transverseMomentum(), m_cfg.ptMin, m_cfg.ptMax) and
+    return within(pT, m_cfg.ptMin, m_cfg.ptMax) and
```
Comment (paulgessinger) [inline Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/TrackParameterSelector.cpp:47]: Unrelated change?

File: Examples/Io/Csv/src/CsvTrackParameterReader.cpp [hunk 12]
```diff
@@ -106,7 +104,9 @@ ActsExamples::ProcessCode ActsExamples::CsvTrackParameterReader::read(
     cov(Acts::eBoundQOverP, Acts::eBoundPhi) = d.cov_qopphi;
     cov(Acts::eBoundQOverP, Acts::eBoundTheta) = d.cov_qoptheta;
 
-    trackParameters.emplace_back(surface, params, q, cov);
+    // TODO we do not have a hypothesis at hand here. defaulting to pion
+    trackParameters.emplace_back(surface, params, cov,
+                                 Acts::ParticleHypothesis::pion());
```
Comment (paulgessinger) [inline Examples/Io/Csv/src/CsvTrackParameterReader.cpp:109]: We control the output here, so might be worth considering to change the CSV format to include this.

**Review by paulgessinger — COMMENTED** [review]
Comments mostly on the EDM intersection. Do we think copies of `ParticleHypothesis` will be expensive enough for us to care?

In many cases the hypothesis objects are copied (they're not right now in the Track EDM, but see my comments on that).

There's a number of (ostensibly at least) unrelated changes included as well.

File: Core/include/Acts/EventData/TrackProxy.hpp [hunk 13]
```diff
@@ -362,10 +357,31 @@ class TrackProxy {
     return parameters()[eBoundQOverP];
   }
 
+  /// Get the particle hypothesis
+  /// @return the particle hypothesis
+  const ParticleHypothesis& particleHypothesis() const {
```
[the remark this answers is not in the document]
Comment (benjaminhuth) [reply Core/include/Acts/EventData/TrackProxy.hpp:362]:
```suggestion
  ParticleHypothesis particleHypothesis() const {
```
mind the reference

[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/TrackProxy.hpp:362]: Indeed, updated my comment.

File: Fatras/include/ActsFatras/EventData/Particle.hpp [hunk 14]
```diff
@@ -163,10 +164,13 @@ class Particle {
   /// Particle mass.
   constexpr Scalar mass() const { return m_mass; }
 
+  /// Particle hypothesis.
+  constexpr Acts::ParticleHypothesis hypothesis() const {
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Fatras/include/ActsFatras/EventData/Particle.hpp:168]: in this implementation the `Particle` is more general than the `ParticleHypothesis` as we strip off the sign of pdg and charge. We could add a `hypothesis` member to avoid the creation of temporaries but otherwise I do not see how to embed this further

**Comment by andiwand** [comment]
> Comments mostly on the EDM intersection. Do we think copies of `ParticleHypothesis` will be expensive enough for us to care?
> 
> In many cases the hypothesis objects are copied (they're not right now in the Track EDM, but see my comments on that).

This is what we discussed once. I guess it is still small enough to not have a huge impact but that would need some testing. The alternative is to make the hypothesis singleton and then just point to it. We only need a few instances anyways I guess

> There's a number of (ostensibly at least) unrelated changes included as well.

Should be gone now

**Comment by benjaminhuth** [comment]
> > Comments mostly on the EDM intersection. Do we think copies of `ParticleHypothesis` will be expensive enough for us to care?
> > In many cases the hypothesis objects are copied (they're not right now in the Track EDM, but see my comments on that).
> 
> This is what we discussed once. I guess it is still small enough to not have a huge impact but that would need some testing. The alternative is to make the hypothesis singleton and then just point to it. We only need a few instances anyways I guess

So the footprint of a particle hypothesis for singly charged particles is 64 bit I think (one `int32_t` for pdg, one `float` for mass).
I think pointer access with singletons or so would be substantially slower.

**Comment by andiwand** [comment]
this is green now - could you have a second look @paulgessinger / @benjaminhuth ?

File: Core/src/Material/SurfaceMaterialMapper.cpp (hunk 8 above)
Comment (paulgessinger) [reply Core/src/Material/SurfaceMaterialMapper.cpp:238]: Right, I see, but I'm wondering if we specifically need to be able to encode Geantinos to hand them over to G4 for the material mapping.

Comment (andiwand) [reply Core/src/Material/SurfaceMaterialMapper.cpp:238]: For me it depends on if we want to use a Geantino hypothesis during reconstruction or not because currently the `ParticleHypothesis` is only used for reconstruction and I think that is a nice separation we should keep.

If we want to add Geantinos we would need some special PDG values or we need a variant to map this correctly

Comment (andiwand) [reply Core/src/Material/SurfaceMaterialMapper.cpp:238]: :+1: I added a geantino for this use case

**Review by benjaminhuth — APPROVED** [review]
Okay, I think this is in a very good shape now. Should go in as soon as we target v30.

**Comment by paulgessinger** [comment]
This has conflicts in the root hashes now.

File: Core/include/Acts/EventData/TrackProxy.hpp [hunk 15]
```diff
@@ -371,6 +371,14 @@ class TrackProxy {
     return component<ParticleHypothesis, hashString("particleHypothesis")>();
   }
 
+  /// Set a new particle hypothesis for this track
+  /// @param particleHypothesis The particel hypothesis to set
```
Comment (paulgessinger) [inline Core/include/Acts/EventData/TrackProxy.hpp:375]:
```suggestion
  /// @param particleHypothesis The particle hypothesis to set
```

File: Core/include/Acts/EventData/TrackProxy.hpp [hunk 16]
```diff
@@ -362,10 +357,31 @@ class TrackProxy {
...
+  /// @return the particle hypothesis
+  const ParticleHypothesis& particleHypothesis() const {
+    return component<ParticleHypothesis, hashString("particleHypothesis")>();
+  }
+
+  /// Get the particle hypothesis
+  /// Mutable version
+  /// @return the particle hypothesis
+  template <bool RO = ReadOnly, typename = std::enable_if_t<!RO>>
+  ParticleHypothesis& particleHypothesis() {
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/TrackProxy.hpp:370]: I think this needs to actually go away: the backend needs to be able to create these on the fly.

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp [hunk 17]
```diff
@@ -79,33 +79,45 @@ struct WeightedComponentReducerLoop {
 
   template <typename stepper_state_t>
   static ActsScalar absoluteMomentum(const stepper_state_t& s) {
+    std::cout << "gsf absoluteMomentum " << s.particleHypothesis.mass() << " "
+              << s.particleHypothesis.absoluteCharge() << " "
+              << s.particleHypothesis.absolutePdg() << std::endl;
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:84]: Still there from debugging?

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp [hunk 18]
```diff
@@ -79,33 +79,45 @@ struct WeightedComponentReducerLoop {
...
+          return sum + cmp.weight * s.particleHypothesis.extractMomentum(
+                                        cmp.state.pars[eFreeQOverP]);
         });
   }
 
   template <typename stepper_state_t>
   static Vector3 momentum(const stepper_state_t& s) {
+    std::cout << "gsf momentum " << s.particleHypothesis.mass() << " "
+              << s.particleHypothesis.absoluteCharge() << " "
+              << s.particleHypothesis.absolutePdg() << std::endl;
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:98]: And here.

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp [hunk 19]
```diff
@@ -79,33 +79,45 @@ struct WeightedComponentReducerLoop {
...
+                               cmp.state.pars[eFreeQOverP]) *
                            cmp.state.pars.template segment<3>(eFreeDir0);
         });
   }
 
   template <typename stepper_state_t>
   static ActsScalar charge(const stepper_state_t& s) {
+    std::cout << "gsf charge " << s.particleHypothesis.mass() << " "
+              << s.particleHypothesis.absoluteCharge() << " "
+              << s.particleHypothesis.absolutePdg() << std::endl;
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:114]: And here.

**Review by paulgessinger — APPROVED** [review]
Let's go!
